chore: add apigateway README file (#430)
DataDog/datadog-cdk-constructs: src/apigateway/README.md

     },
   });
   ```

### Supported versions and web frameworks

| Runtime | Datadog Tracer | Tracer version       | Frameworks                                                                                                                            |
| ------- | -------------- | -------------------- | ------------------------------------------------------------------------------------------------------------------------------------- |
| Node.js | `dd-trace-js`  | v4.50.0+ or v5.26.0+ | express, fastify, hapi, koa, microgateway-core, next, paperplane, restify, router, apollo                                             |
| Go      | `dd-trace-go`  | v1.72.1+             | chi (v5), httptreemux (v5), echo (v3-v4), go-restful (v3), fiber (v2), gin, gorilla/mux, httprouter, fasthttp (v1), goji (v1), gqlgen |
| Python  | `dd-trace-py`  | v3.1.0+              | aiohttp, asgi, bottle, cherrypy, django/djangorestframework, falcon, fastapi, flask, molten, pyramid, sanic, starlette, tornado, wsgi |

## API Gateway v1 (REST APIs)

There are three ways to implement API Gateway instrumentation:

### 1. RestApi parameters (recommended)

> Add Datadog instrumentation headers to your API Gateway request parameters. These parameters are applied to every resource.

Reviewer: ```suggestion
Add Datadog instrumentation headers to your API Gateway request parameters. These parameters apply to every resource.
```

  type: apigateway.IntegrationType.HTTP_PROXY,
  integrationHttpMethod: "ANY",
  options: {
    connectionType: apigateway.ConnectionType.INTERNET,
    requestParameters: DatadogAPIGatewayRequestParameters,
  },
  uri: `http://${loadBalancer.loadBalancerDnsName}`,
});

const api = new apigateway.RestApi(this, "MyApi", {
  restApiName: "my-api-gateway",
  deployOptions: { stageName: "prod" },
  defaultIntegration: ddIntegration, // Datadog instrumentation applied here
});
```

### 3. Per-resource integration

Manually add the instrumentation to each resource when calling `addMethod`. This approach is required for resources with custom integrations.

```typescript
api.root.addMethod("ANY", ddIntegration); // Datadog instrumentation applied here
const books = api.root.addResource("books");
books.addMethod("ANY", ddIntegration); // Datadog instrumentation applied here
const book = books.addResource("{id}");
book.addMethod("ANY", ddIntegration); // Datadog instrumentation applied here
```

## API Gateway v2 (HTTP APIs)

Add Datadog instrumentation by using the `DatadogAPIGatewayV2ParameterMapping` constant, which contains the necessary parameter mappings for tracing.

> <div class="alert alert-info">`context.requestTimeEpoch` on v2 APIs provides only second-level granularity, so the synthetic-span duration is approximate. This does not affect tags, Trace Maps, or span linkage.</div>

Reviewer: ```suggestion
<div class="alert alert-info"> <p><code>context.requestTimeEpoch</code> on v2 APIs provides only second-level granularity, so synthetic span duration is approximate. This does not affect tags, Trace Maps, or span linkage.</p> </div>
```

```typescript
// Datadog integration definition
const ddIntegration = new apigatewayv2_integrations.HttpUrlIntegration("HttpUrlIntegration", "https://example.com", {
  parameterMapping: DatadogAPIGatewayV2ParameterMapping,
});

const httpApi = new apigatewayv2.HttpApi(this, "httpApi", {
  apiName: "httpApiName",
  description: "API Gateway v2 for forwarding requests to example.com",
});

httpApi.addRoutes({
  path: "/{proxy+}",
  methods: [apigatewayv2.HttpMethod.ANY],
  integration: ddIntegration, // Datadog instrumentation applied here
});
httpApi.addRoutes({
  path: "/books",
  methods: [apigatewayv2.HttpMethod.ANY],
  integration: ddIntegration, // Datadog instrumentation applied here
});
httpApi.addRoutes({
  path: "/books/{id}",
  methods: [apigatewayv2.HttpMethod.ANY],
  integration: ddIntegration, // Datadog instrumentation applied here
});
```

> ## Important Notes

Reviewer: ```suggestion
## Tracing behavior and configuration notes
```

1. Head-based sampling is still in effect when using this feature. Any sampling rules in use need to be adjusted for the new root span. The service name must match the API Gateway's name as seen in Datadog.

   Example: If you were using a sampling rule like:

   ```bash
   DD_TRACE_SAMPLING_RULES='[{"service": "pythonapp", "sample_rate": 0.5}]'
   ```

   You will need to update your sampling rules:

   - **Option 1:** Change the `service` value to match your API Gateway's name as it appears in Datadog:
     ```bash
     DD_TRACE_SAMPLING_RULES='[{"service": "my-api-gateway", "sample_rate": 0.5}]'
     ```
   - **Option 2:** Remove the `service` key to apply the rule to all root spans:
     ```bash
     DD_TRACE_SAMPLING_RULES='[{"sample_rate": 0.5}]'
     ```

>    Any sampling rules based on `resource_name` for the original web application now need to be updated for the resource name of the API Gateway service.

Reviewer: Vale doesn't like "now".
```suggestion
   Any sampling rules based on `resource_name` for the original web application must be updated for the resource name of the API Gateway service.
```

> 1. The instrumentation adds several headers to track:

Reviewer: ```suggestion
1. The instrumentation adds headers to track the following:
```

## Overview

API Gateway synthetic spans allow you to track and monitor API requests through your API Gateway endpoints. This instrumentation adds necessary headers to your API Gateway integrations, enabling Datadog to properly trace and monitor your API traffic.

## Prerequisites

1. Your application must be running a supported web framework (see list below).

2. Set `DD_TRACE_INFERRED_PROXY_SERVICES_ENABLED=true` in your application environment. If you're using the Datadog ECS Fargate construct, you can enable this by setting the `traceInferredProxyServices` property in the APM configuration:

```typescript
const ecsDatadog = new DatadogECSFargate({
  // ... other configuration
  apm: {
    isEnabled: true,
    traceInferredProxyServices: true, // This sets DD_TRACE_INFERRED_PROXY_SERVICES_ENABLED=true
  },
});
```

3. Use a supported version of the Datadog tracer:

> | Tracer          | Minimum Version |
> | --------------- | --------------- |
> | dd-trace-js 5.x | 5.26.0          |
> | dd-trace-js 4.x | 4.50.0          |
> | dd-trace-go     | 1.72.1          |
> | dd-trace-py     | 3.1.0           |

Reviewer: ```suggestion
### Supported versions and web frameworks

| Runtime | Datadog Tracer | Tracer version | Frameworks |
| ------- | -------------- | ---------------| ---------- |
| Node.js | `dd-trace-js` | v4.50.0+ or v5.26.0+ | express, fastify, hapi, koa, microgateway-core, next, paperplane, restify, router, apollo |
| Go | `dd-trace-go` | v1.72.1+ | chi (v5), httptreemux (v5), echo (v3-v4), go-restful (v3), fiber (v2), gin, gorilla/mux, httprouter, fasthttp (v1), goji (v1), gqlgen |
| Python | `dd-trace-py` | v3.1.0+ | aiohttp, asgi, bottle, cherrypy, django/djangorestframework, falcon, fastapi, flask, molten, pyramid, sanic, starlette, tornado, wsgi |
```
